lib/Optimizer/Transforms/Decomposition.cpp

```diff
@@ -0,0 +1,106 @@
+/*************************************************************** -*- C++ -*- ***
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ *******************************************************************************/
+
+#include "DecompositionPatterns.h"
+#include "cudaq/Frontend/nvqpp/AttributeNames.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeDialect.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Threading.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Generated logic
+//===----------------------------------------------------------------------===//
+namespace cudaq::opt {
+#define GEN_PASS_DEF_DECOMPOSITIONPASS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Pass implementation
+//===----------------------------------------------------------------------===//
+
+struct Decomposition
+    : public cudaq::opt::impl::DecompositionPassBase<Decomposition> {
+  using DecompositionPassBase::DecompositionPassBase;
+
+  /// Initialize the decomposer by building the set of patterns used during
+  /// execution.
+  LogicalResult initialize(MLIRContext *context) override {
+    RewritePatternSet owningPatterns(context);
+    cudaq::populateWithAllDecompositionPatterns(owningPatterns);
+    patterns = FrozenRewritePatternSet(std::move(owningPatterns),
+                                       disabledPatterns, enabledPatterns);
+    return success();
+  }
+
+  void runOnOperation() override {
+    auto module = getOperation();
+
+    // First, we walk the whole module in search for controlled `quake.apply`
+    // operations: If present, we conservatively don't do any decompostions. We
+    // also collect quantum kernels.
+    //
+    // TODO: Evaluate if preventing decompostion when there is at least one
```

Comment:
How would you propose we handle `control-kernel` in the future?

Reply:
We would need to handle it before applying decompositions because of global phase to relative phase issues. We could explore the idea of making patterns that are equivalent up to a global phase to their original operation add a global phase operation that would later become controlled.